### DISCUSSION TITLE
Add a configuration that can enable/disable block comment processing

### DISCUSSION THREAD

**Discussion by paul8263** [spring-projects/spring-shell#492] [General]
Affect version: Spring-shell 1.2.0.RELEASE

In 1.2.0.RELEASE spring shell tends to erase everything between block command quotes. 
However sometime we do not need this feature as we might need to pass a URL regex like '\*/\*/\*'. If doing so, spring shell will keep characters before last '/\*' and after last '\*/'. As a result we get '\*/\*\*', which is buggy.

I suggest we had better introduce a configuration that can enable/disable block comment processing.

**Comment by jvalkeal** [comment]
`1.2.0.RELEASE` is 6 years old release so can you try `2.1.0`.

To you question I have a hard time to understand what you mean. Is this about command file? Can you show a real example what happens with a shell.

**Comment by paul8263** [comment]
Hi @jvalkeal ,

Thank you very much for your reply.

It is about command line param. As shown in the example below:
```java
public class MainShell {
    public static void main(String[] args) throws Exception {
        Bootstrap bootstrap = new Bootstrap();
        JLineShellComponent shell = bootstrap.getJLineShellComponent();
        shell.executeCommand("command --str */*/*");
    }
}
```

```java
@Component
public class MyCommand implements CommandMarker {
    @CliCommand(value = "command")
    public String execCommand(
            @CliOption(key = {"str"}) String str
    ) throws Exception {
        return str;
    }
}
```

The problem is, when we pass `*/*/*` as the parameter 'str' but actually we get `*/**`。
It seems this feature cannot be turned off . The relevant code block is:

AbstractShell::executeCommand:
```java
// We support simple block comments; ie a single pair per line
if (!inBlockComment && line.contains("/*") && line.contains("*/")) {
    blockCommentBegin();
    String lhs = line.substring(0, line.lastIndexOf("/*"));
    if (line.contains("*/")) {
        line = lhs + line.substring(line.lastIndexOf("*/") + 2);
        blockCommentFinish();
    } else {
        line = lhs;
    }
}
if (inBlockComment) {
    if (!line.contains("*/")) {
        return new CommandResult(true);
    }
    blockCommentFinish();
    line = line.substring(line.lastIndexOf("*/") + 2);
}
// We also support inline comments (but only at start of line, otherwise valid
// command options like http://www.helloworld.com will fail as per ROO-517)
if (!inBlockComment && (line.trim().startsWith("//") || line.trim().startsWith("#"))) { // # support in ROO-1116
    line = "";
}
```

Also I got two more questions and need some help:
1. Can spring shell 2.1.0 temporarily turn off this feature by centain configuration?
2. It seems that 2.1.0 version works with spring boot, but we prefer a light-weight implementation. I prefer to upgrade but have to take it into consideration.

**Comment by jvalkeal** [comment]
This is from hoodie right? Unfortunately you are on an unsupported stack here as this is such an old code. Part of that code in shell 1.x code originates from spring roo(which means we're talking 10+ years). Spring Boot then became later which these days is the bootstrapping framework.

If you need to fix something on spring-shell 1.x I think only option is to fork it and make needed quick fixes to give more time for possible rework/overhaul on top of more modern spring-shell.

Since early days of spring shell 1.x it has been rewritten two times so very little is in place these days. What comes for your question 1, it's not valid as offending code is not there anymore.

About your question 2; Spring shell 1.x used spring-framework as spring-boot didn't exist at that time, thus the `Bootstrap` you have. You'd be surprised how good solution spring-shell is with spring-boot as there's been a years of maintenance in a framework itself.

Also if you choose to migrate, you have a good path forward to keep dependencies up-to-date, get a change to be on a supported versions from both spring and jdk.

**Comment by jvalkeal** [comment]
I think it's better to move this into discussion.

**Reply by paul8263** [reply]
Hi @jvalkeal ,
Thank you for your reply. 
Currently I am waiting for the Hudi community's decision about whether upgrading Spring shell or fixing 1.2.0.RELEASE.

**Comment by paul8263** [comment]
Hi @jvalkeal ,

I got a problem and need your help.

How could I run the shell command via Java code in 2.1.1 like those in 1.2.1.RELEASE:
```java
Bootstrap bootstrap = new Bootstrap();
JLineShellComponent shell = bootstrap.getJLineShellComponent();
shell.executeCommand("help");
```

We need to upgrade the unit tests.

Thank you very much.

**Reply by jvalkeal** [reply]
Are you asking if we have something like https://www.baeldung.com/run-shell-command-in-java build-in?

**Reply by paul8263** [reply]
No. What I mean is how to execute and test a shell method defined with @ShellMethod using a command line.
For example:
In unit test, we execute `Object result = xxx.executeCommand("add 1 2")` and actually the underlying method can be tested.
```java
@ShellMethod("Add two integers together.")
public int add(int a, int b) {
    return a + b;
}
```
Thank you for your reply.

**Reply by jvalkeal** [reply]
Testing is difficult what comes for integration tests as there's console and jline layers in a middle. Plan is to make something better in #489 but in a meanwhile we have a relatively good tests in spring-shell itself.

https://github.com/spring-projects/spring-shell/blob/main/spring-shell-core/src/test/java/org/springframework/shell/component/StringInputTests.java shows some ideas to use dump terminal(from plain jvm you don't have working pty). JLine is using threads so you need to wait output and things gets awkward if ansi characters are involved.

We've been playing with ideas around testing shell app as a black box using node based e2e tests in https://github.com/spring-projects/spring-shell/tree/main/e2e. It works remarkable well as node has a pty libs(so you get a working os level console env and JLine doesn't fallback to dump terminal) and it feed stuff into xterm.js(used by all JS based ide's and online tools) which gives us working solution for ansi escape characters(you get what is actually rendered on screen). It also allows to test shell app both fatjar and graalvm binary.

**Reply by paul8263** [reply]
Hi @jvalkeal ,
Sorry for my late response.

I found a way to test the spring ShellMethod in Java, the example code is like:
```java
@RunWith(SpringRunner.class)
@SpringBootTest(properties = {"spring.shell.interactive.enabled=false", "spring.shell.command.script.enabled=false"})
public class DemoTest {
    @Autowired
    private Shell shell;

    @Test
    public void testDemo() {
        Object result = shell.evaluate(() -> "xxxCommand --sql 'select * from t1'");
    }
}
```
However it seems that `shell.evaluate()` has some problems when dealing with parameter value with "blank space" in it, even though we use quotes or double quotes. Please see the example code above.

Conversely, the command: xxxCommand --sql 'select * from t1' works correctly when run inside a terminal(started via java -jar xxx.jar). The value of the parameter sql is 'select * from t1'.

**Reply by paul8263** [reply]
Finally I found a solution that we can implement the Input interface, override the words method with the correct commandline splitting logics and pass it to `shell.evaluate()`.

**Reply by paul8263** [reply]
There might be another problem that a shell method with an Integer type patameter cannot receive a negative number.
